src/main/java/com/walmartlabs/x12/util/split/X12TransactionSplitter.java

```diff
@@ -0,0 +1,388 @@
+package com.walmartlabs.x12.util.split;
+
+import com.walmartlabs.x12.SegmentIterator;
+import com.walmartlabs.x12.X12Segment;
+import com.walmartlabs.x12.X12TransactionSet;
+import com.walmartlabs.x12.exceptions.X12ErrorDetail;
+import com.walmartlabs.x12.exceptions.X12ParserException;
+import com.walmartlabs.x12.rule.X12Rule;
+import com.walmartlabs.x12.standard.StandardX12Parser;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * 
+ * Given an EDI message 
+ * Break it up into multiple messages based on the ST/SE transaction
+ * 
+ * This has bare-bones validation
+ * 
+ * Each transaction will have the following:
+ * - the ISA header
+ * - the GS header it is part of
+ * - the ST/SE transaction
+ * - the GE trailer it is part of
+ * - the ISE trailer
+ *
+ */
+public class X12TransactionSplitter {
+    
+    private static final String EOL = "\r\n";
+    
+    private List<X12Rule> rules;
+    
+    /**
+     * split the EDI message (raw file)
+     * so that each transaction 
+     * is a separate "document" w/ the original 
+     * ISA and group headers and trailers
+     * 
+     * @param sourceData
+     * @return the list of EDI transactions
+     * @throws X12ParserException
+     */
+    public List<String>  split(String sourceData) {
+        if (StringUtils.isEmpty(sourceData)) {
+            return Collections.emptyList();
+        } else {
+            List<X12Segment> segmentList = new StandardX12Parser().splitSourceDataIntoSegments(sourceData.trim());
+            return this.split(segmentList);
+        }
+    }
+    
+    /**
+     * split the EDI message (list of X12Segments)
+     * so that each transaction 
+     * is a separate "document" w/ the original 
+     * ISA and group headers and trailers
+     * 
+     * @param segmentList 
+     * @return the list of EDI transactions
+     * @throws X12ParserException
+     */
+    public List<String> split(List<X12Segment> segmentList) {
+        if (CollectionUtils.isEmpty(segmentList)) {
+            return Collections.emptyList();
+        } else {
+            // do some basic validation
+            // on the entire EDI message
+            // using X12Rule set provided
+            this.runRules(segmentList);
+            
+            // parse and split
+            // the EDI message
+            List<TransactionHolder> transactionHolders = this.doParse(segmentList);
+            return this.doSplit(transactionHolders);
+        }
+    }
+    
+    /**
+     * add an {@link X12Rule} to the splitter
+     * @param rule
+     */
+    public void registerX12Rule(X12Rule rule) {
+        if (this.rules == null) {
+            this.rules = new ArrayList<>();
+        }
+        
+        this.rules.add(rule);
```

Comment:
Should we do null check on `rule`

Reply:
added the null check during the running...